File: src/main/java/net/lomeli/trophyslots/core/command/CommandUnlockSlots.java
package net.lomeli.trophyslots.core.command;

import net.minecraft.command.CommandBase;
import net.minecraft.command.ICommandSender;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.util.ChatComponentText;
import net.minecraft.util.ChatComponentTranslation;
import net.minecraft.util.StatCollector;

import net.lomeli.trophyslots.TrophySlots;
import net.lomeli.trophyslots.core.SlotUtil;
import net.lomeli.trophyslots.core.network.MessageSlotsClient;

public class CommandUnlockSlots extends CommandBase {
    @Override
    public String getCommandName() {
        return "unlock-slots";
    }

    @Override
    public int getRequiredPermissionLevel() {
        return 2;
    }

    @Override
    public String getCommandUsage(ICommandSender sender) {
        return "command.trophyslots.unlock-slots.usage";
    }

    @Override
    public void processCommand(ICommandSender sender, String[] args) {
        if (args.length == 2 || args.length == 3) {
            EntityPlayerMP player = args.length == 2 ? getCommandSenderAsPlayer(sender) : getPlayer(sender, args[1]);
            int newSlots = args.length == 2 ? parseString(args[1]) : parseString(args[2]);
            if (player != null && (newSlots > 0 && newSlots <= SlotUtil.getMaxSlots())) {
                if (SlotUtil.hasUnlockedAllSlots(player))
                    sender.addChatMessage(new ChatComponentText(String.format(StatCollector.translateToLocal("command.trophyslots.unlock-slots.error.all"), player.getDisplayName())));
                else {
                    int slots = SlotUtil.getSlotsUnlocked(player);
                    slots += newSlots;
                    if (slots >= SlotUtil.getMaxSlots())
                        TrophySlots.proxy.unlockAllSlots(player);
                    else {
                        SlotUtil.setSlotsUnlocked(player, slots);
                        TrophySlots.packetHandler.sendTo(new MessageSlotsClient(slots), player);
                        player.addChatMessage(new ChatComponentText(String.format(StatCollector.translateToLocal("msg.trophyslots.unlockSlot"), newSlots)));
                    }
                    sender.addChatMessage(new ChatComponentText(String.format(StatCollector.translateToLocal("command.trophyslots.unlock-slots.success"), newSlots, player.getDisplayName())));
                }
            } else
                sender.addChatMessage(new ChatComponentText(String.format(StatCollector.translateToLocal("command.trophyslots.unlock-slots.error"), SlotUtil.getMaxSlots())));
        } else
            sender.addChatMessage(new ChatComponentTranslation(getCommandUsage(sender)));
    }

    private int parseString(String st) {
        try {
            return Integer.parseInt(st);
        } catch (Exception e) {
            return -1;
        }
    }
}

File: src/main/kotlin/net/lomeli/trophyslots/core/command/CommandUnlockSlots.kt
package net.lomeli.trophyslots.core.command

import net.lomeli.trophyslots.capabilities.slots.SlotManager
import net.minecraft.command.CommandBase
import net.minecraft.command.ICommandSender
import net.minecraft.entity.player.EntityPlayerMP
import net.minecraft.server.MinecraftServer
import net.minecraft.util.text.TextComponentString
import net.minecraft.util.text.TextComponentTranslation
import net.minecraft.util.text.translation.I18n

class CommandUnlockSlots : CommandBase() {
    override fun execute(server: MinecraftServer?, sender: ICommandSender?, args: Array<out String>?) {
        if (sender != null) {
            if (args != null && (args.size == 2 || args.size == 3)) {
                val player: EntityPlayerMP
                val newSlots: Int
                if (args.size == 3) {
                    newSlots = parseInt(args[2])
                    player = CommandBase.getPlayer(server, sender, args[1])
                } else {
                    newSlots = parseInt(args[1])
                    player = CommandBase.getCommandSenderAsPlayer(sender)
                }
                val slotInfo = SlotManager.getPlayerSlotInfo(player)!!
                if (player != null && (newSlots > 0 && newSlots <= slotInfo.getMaxSlots())) {
                    if (slotInfo.isAtMaxSlots())
                        sender.sendMessage(TextComponentString(I18n.translateToLocal("command.trophyslots.unlock_slots.error.all").format(player.displayName.unformattedText)))
                    else {
                        val slots = slotInfo.getSlotsUnlocked() + newSlots
                        if (slots >= slotInfo.getMaxSlots()) {
                            slotInfo.setSlots(slotInfo.getMaxSlots())
                            SlotManager.updateClient(player, slotInfo)
                            player.sendMessage(TextComponentString(I18n.translateToLocal("msg.trophyslots.unlock_all")))
                        } else {
                            slotInfo.setSlots(slots)
                            SlotManager.updateClient(player, slotInfo)
                            player.sendMessage(TextComponentString(I18n.translateToLocal("msg.trophyslots.unlock_slot").format(newSlots)))
                        }
                        sender.sendMessage(TextComponentString(I18n.translateToLocal("command.trophyslots.unlock_slots.success").format(newSlots, player.displayName.unformattedText)))
                    }
                } else
                    sender.sendMessage(TextComponentString(I18n.translateToLocal("command.trophyslots.unlock_slots.error").format(slotInfo.getMaxSlots())))
            } else
                sender.sendMessage(TextComponentTranslation(getUsage(sender)))
        }
    }

    override fun getName(): String? = "unlock-slots"

    override fun getUsage(sender: ICommandSender?): String? = "command.trophyslots.unlock_slots.usage"

    override fun getRequiredPermissionLevel(): Int = 2
}